File: d1/solve.go
package d1

import (
	"bufio"
	"io"

	"github.com/itsbth/aoc-2023/runner"
)

var MAP = []string{
	"one",
	"two",
	"three",
	"four",
	"five",
	"six",
	"seven",
	"eight",
	"nine",
}

var m = matcher{}

func sum(line string) (uint64, uint64) {
	var sum1, sum2 uint64

	found := false

	m.reset()

	// forward scan, find first ASCII digit and first spelled-out digit
outer:
	for i := 0; i < len(line); i++ {
		if line[i] >= '0' && line[i] <= '9' {
			val := uint64(line[i]-'0') * 10
			sum1 += val
			if !found {
				sum2 += val
			}
			break outer
		} else if !found && line[i] >= 'a' && line[i] <= 'z' {
			if v, ok := m.match(line[i]); ok {
				val := uint64(v) * 10
				sum2 += val
				found = true
			}
		}
	}

	// reverse scan, find last ASCII digit and last spelled-out digit
	found = false
	m.resetBW()
outer2:
	for i := len(line) - 1; i >= 0; i-- {
		if line[i] >= '0' && line[i] <= '9' {
			val := uint64(line[i] - '0')
			sum1 += val
			if !found {
				sum2 += val
			}
			break outer2
		} else if !found && line[i] >= 'a' && line[i] <= 'z' {
			if v, ok := m.matchBW(line[i]); ok {
				val := uint64(v)
				sum2 += val
				found = true
			}
		}
	}

	return sum1, sum2
}

type solver struct{}

var _ runner.Solver = solver{}

func (solver) Solve(input io.Reader) (int, int, error) {
	total1, total2 := uint64(0), uint64(0)

	scanner := bufio.NewScanner(input)
	for scanner.Scan() {
		line := scanner.Text()
		c1, c2 := sum(line)
		total1 += c1
		total2 += c2
	}

	return int(total1), int(total2), nil
}

func init() {
	runner.Register(2023, 1, solver{}, "default")
}

File: d1/solve.py
import re

# digits only
P1 = re.compile(r'(\d)')

# digits or spelled out numbers
# use lookahead to match overlapping numbers
P2 = re.compile(r'(?=(\d|one|two|three|four|five|six|seven|eight|nine))')

VALUE = {
    'one': 1,
    'two': 2,
    'three': 3,
    'four': 4,
    'five': 5,
    'six': 6,
    'seven': 7,
    'eight': 8,
    'nine': 9
}

def expand(s):
    if s in VALUE:
        return VALUE[s]
    else:
        return int(s)

def solve(s):
    s1 = 0
    s2 = 0
    for line in s.splitlines():
        do = P1.findall(line)
        ds = P2.findall(line)
        # debug: print line s1 (first and last digit) and s2 (first and last digit/number)
        # print(f"{line} {do[0]}{do[-1]} {expand(ds[0])}{expand(ds[-1])}")
        s1 += int(do[0]) * 10 + int(do[-1])
        s2 += expand(ds[0]) * 10 + expand(ds[-1])
    print(f"Part 1: {s1}")
    print(f"Part 2: {s2}")


if __name__ == '__main__':
    f = open('INPUT', 'r')
    solve(f.read())
    f.close()